drivers/flash/flash_stm32f7x.c

```diff
@@ -0,0 +1,201 @@
+
+	return rc;
+}
+
+int flash_stm32_block_erase_loop(struct device *dev, unsigned int offset,
+				 unsigned int len)
+{
+	struct flash_pages_info info;
+	u32_t start_sector, end_sector;
+	u32_t i;
+	int rc = 0;
+
+	rc = flash_get_page_info_by_offs(dev, offset, &info);
+	if (rc) {
+		return rc;
+	}
+	start_sector = info.index;
+	rc = flash_get_page_info_by_offs(dev, offset + len - 1, &info);
+	if (rc) {
+		return rc;
+	}
+	end_sector = info.index;
+
+	for (i = start_sector; i <= end_sector; i++) {
+		rc = erase_sector(dev, i);
+		if (rc < 0) {
+			break;
+		}
+	}
+
+	return rc;
+}
+
+int flash_stm32_write_range(struct device *dev, unsigned int offset,
+			    const void *data, unsigned int len)
+{
+	int i, rc = 0;
+
+	for (i = 0; i < len; i++, offset++) {
+		rc = write_byte(dev, offset, ((const u8_t *) data)[i]);
+		if (rc < 0) {
+			return rc;
+		}
+	}
+
+	return rc;
+}
+
+/*
+ * Different SoC flash layouts are specified in across various
+ * reference manuals, but the flash layout for a given number of
+ * sectors is consistent across these manuals, with one "gotcha". The
+ * number of sectors is given by the HAL as FLASH_SECTOR_TOTAL.
+ *
+ * The only "gotcha" is that when there are 24 sectors, they are split
+ * across 2 "banks" of 12 sectors each, with another set of small
+ * sectors (16 KB) in the second bank occurring after the large ones
+ * (128 KB) in the first. We could consider supporting this as two
+ * devices to make the layout cleaner, but this will do for now.
+ */
+#ifndef FLASH_SECTOR_TOTAL
+#error "Unknown flash layout"
+#else  /* defined(FLASH_SECTOR_TOTAL) */
+#if FLASH_SECTOR_TOTAL == 8
+static const struct flash_pages_layout stm32f7_flash_layout[] = {
+	/*
+	 * RM0410, table 5
+	 */
+	{.pages_count = 4, .pages_size = KB(32)},
+	{.pages_count = 1, .pages_size = KB(128)},
+	{.pages_count = 3, .pages_size = KB(256)},
+};
+#elif FLASH_SECTOR_TOTAL == 12
+static const struct flash_pages_layout stm32f7_flash_layout[] = {
+	/*
+	 * RM0410, table 3
+	 */
+	{.pages_count = 4, .pages_size = KB(32)},
+	{.pages_count = 1, .pages_size = KB(128)},
+	{.pages_count = 7, .pages_size = KB(256)},
+};
+#elif FLASH_SECTOR_TOTAL == 16
+static const struct flash_pages_layout stm32f7_flash_layout[] = {
+	/* 
+	 * RM0410, table 6
+	 */
+	{.pages_count = 4, .pages_size = KB(16)},
+	{.pages_count = 1, .pages_size = KB(64)},
+	{.pages_count = 3, .pages_size = KB(128)},
+	{.pages_count = 4, .pages_size = KB(16)},
+        {.pages_count = 1, .pages_size = KB(64)},
+        {.pages_count = 3, .pages_size = KB(128)},
+
+};
+#elif FLASH_SECTOR_TOTAL == 24
+static const struct flash_pages_layout stm32f7_flash_layout[] = {
+	/*
+	 * RM0410, table 4
+	 */
+	{.pages_count = 4, .pages_size = KB(16)},
+	{.pages_count = 1, .pages_size = KB(64)},
+	{.pages_count = 7, .pages_size = KB(128)},
+	{.pages_count = 4, .pages_size = KB(16)},
+	{.pages_count = 1, .pages_size = KB(64)},
+	{.pages_count = 7, .pages_size = KB(128)},
+};
```

Comment:
As said above, there are two cases depending of the `nDBANK` bit of the `FLASH_OPTCR` register. You could have a look how I have done that there: https://github.com/aurel32/zephyr/commit/7b6e90dd019dd68cee463c40230f23aa0fdbd81b#diff-322955a6f2ebc21dba22c2c642a8b3faR129